File: core/src/main/java/com/pluxity/station/StationLineService.java
package com.pluxity.station;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.util.CollectionUtils;

@Service
@RequiredArgsConstructor
public class StationLineService {

    private final StationLineRepository stationLineRepository;

    @Transactional
    public void save(Station station, Line line) {
        stationLineRepository.save(StationLine.builder().station(station).line(line).build());
    }

    @Transactional(readOnly = true)
    public List<Long> findLinesByStation(Station station) {
        return stationLineRepository.findByStationOrderByCreatedAtDesc(station).stream()
                .map(v -> v.getLine().getId())
                .toList();
    }

    @Transactional
    public void deleteByStation(Station station) {
        stationLineRepository.deleteByStation(station);
    }

    @Transactional(readOnly = true)
    public Map<Station, List<Long>> findLineMapByStationIds(List<Station> stations) {
        if (CollectionUtils.isEmpty(stations)) {
            return Collections.emptyMap();
        }
        return stationLineRepository.findByStationInOrderByCreatedAtDesc(stations).stream()
                .collect(
                        Collectors.groupingBy(
                                StationLine::getStation,
                                Collectors.mapping(v -> v.getLine().getId(), Collectors.toList())));
    }

    @Transactional(readOnly = true)
    public boolean checkAlreadyConnect(Station station, Line line) {
        return stationLineRepository.existsByStationAndLine(station, line);
    }

    @Transactional
    public void deleteStationLine(Station station, Line line) {
        stationLineRepository.deleteByStationAndLine(station, line);
    }
}

File: core/src/main/kotlin/com/pluxity/station/StationLineService.kt
package com.pluxity.station

import com.pluxity.global.constant.ErrorCode
import com.pluxity.global.exception.CustomException
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import org.springframework.util.CollectionUtils

@Service
class StationLineService(
    private val stationLineRepository: StationLineRepository,
) {
    @Transactional
    fun save(
        station: Station,
        line: Line,
    ) {
        val stationLine = stationLineRepository.save(StationLine(station = station, line = line))
        line.addStationLine(stationLine)
    }

    @Transactional(readOnly = true)
    fun findLinesByStation(station: Station): List<Long> =
        stationLineRepository.findByStationOrderByCreatedAtDesc(station).map { it.line.requiredId }

    @Transactional
    fun deleteByStation(station: Station) {
        stationLineRepository.deleteByStation(station)
    }

    @Transactional(readOnly = true)
    fun findLineMapByStationIds(stations: List<Station>): Map<Station, List<Long>> {
        if (CollectionUtils.isEmpty(stations)) {
            return mapOf()
        }
        return stationLineRepository
            .findByStationInOrderByCreatedAtDesc(stations)
            .groupBy { it.station }
            .mapValues { (_, lines) -> lines.map { it.line.requiredId } }
    }

    @Transactional(readOnly = true)
    fun checkAlreadyConnect(
        station: Station,
        line: Line,
    ): Boolean = stationLineRepository.existsByStationAndLine(station, line)

    @Transactional
    fun deleteStationLine(
        station: Station,
        line: Line,
    ) {
        val stationLine: StationLine =
            stationLineRepository
                .findByStationAndLine(station, line)
                ?: throw CustomException(ErrorCode.NOT_FOUND_STATION_LINE, station.id, line.id)
        line.removeStationLine(stationLine)
        stationLineRepository.delete(stationLine)
    }
}
